### PR TITLE
NIFI-12888 In AbstractEmailProcessor recreate email receiver when token expires.

### PR DESCRIPTION
**Pull request by tpalfy**
# Summary

[NIFI-12888](https://issues.apache.org/jira/browse/NIFI-12888)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
I agree with @dan-s1, the substantive changes are straightforward, but the multiple levels of indentation, particularly with the complex conditionals, make the updated logic more difficult to follow.

File: nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/AbstractEmailProcessor.java [hunk 1]
```diff
@@ -324,6 +330,18 @@ private synchronized void initializeIfNecessary(ProcessContext context, ProcessS
...
+        boolean oauthDetailsRefreshed = this.oauth2AccessTokenProviderOptional.isPresent()
+                &&
+                (
+                        this.oauth2AccessDetails == null
+                                ||
+                                !oauth2AccessDetails.equals(this.oauth2AccessTokenProviderOptional.get().getAccessDetails())
+                );
+
+        return oauthDetailsRefreshed;
+    }
```
[the remark this answers is not in the document]
Comment (tpalfy) [reply nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/AbstractEmailProcessor.java:343]: I usually find it the case that once you understand the complex condition you find the cluttered presentation more appealing. But only after that. When it's cluttered, one needs to look more carefully to understand the relationship between the parts and think through more carefully.

But I think can come up with a compromise and make it more streamlined a bit.

As for doing it in a return statement - I find it much better to return a single variable.

**Review by exceptionfactory — COMMENTED** [review]
Thanks for making the adjustments @tpalfy, the latest version is much more readable.

@dan-s1 has one remaining comment about the naming convention of the new method, so I will defer to his feedback on that remaining point for now.

**Comment by dan-s1** [comment]
@exceptionfactory I think with the latest commit everything seems okay.

**Review by pvillard31 — APPROVED** [review]

**Comment by AnTu2702** [comment]
@tpalfy Hey. How are you?

First of all: Thanks for the fix/workaround in NIFI 1.26.
Although things are basically working there, we still don't seem to be through.

The ConsumeIMAP Processor is continuously logging this ERROR...
(although the mailbox can and will be processed successfully using a refreshed token...)

image

We still have serious doubts to take this into production because of the log being spammed and of course the fact, that we have no clue, what causes this error.

Could you help and double check your fix/workaround?
Thank you in advance...

**Comment by tpalfy** [comment]
Hi @AnTu2702,

Do you happen to try connecting to an Office365 account?
The issue you mention didn't occur when I was testing but when googled it, it gave me several results users complaining about this when using Office365.

According to one search result this can be caused by clients that are trying to connect too frequently. Can you try lowering the schedule period and see if that helps?
Others suggest checking the OAuth config of Office365.

Couple of links for reference:
https://confluence.atlassian.com/jirakb/imap-fails-with-a3-bad-user-is-authenticated-but-not-connected-error-in-jira-server-integrated-with-office365-665420269.html
https://github.com/mscdex/node-imap/issues/689

**Comment by AnTu2702** [comment]
Der alte E-Mail Stack kann abgerissen werden jetzt.
Wir lassen aber den parallelen Betrieb des alten und neuen WhatsApp Postfachs in NIFI vorerst aktiv, bis wieder WhatsApp Emails eintrudeln. Danach schalten wir den Alias um und reißen die alten NIFI Prozesse ab.

**Pull request closed without merging**